Recursive child processes

garunagiri opened the pull request:
Functionality to recursively search for child processes until there are no more, up to a certain depth.
This information is stored into a persistent process tree. Parent links can be looked up by providing a child process' PID and the host the link was run on.

unkempthenry reviewed (approved):
Looks pretty good to me.  One suggestion though -- this stuff can get a little complicated.  Adding some docstrings to these functions (epseically in the `ProcessTree` class) about why each function is there and how it should be used might make it a lot easier for the unitiated to grok.

clenk reviewed (commented):
Reading through this I was confused at first because there was only 1 `ProcessTree` and `pid_to_guids_map` and `guid_to_processnode_map` are dictionaries of dictionaries (i.e. `host_to_pid_to_guids_map` would be more accurate). Would it make sense to have a `ProcessTree` per host so those mappings can be the simpler dictionaries their names imply?

garunagiri commented on app/c_processtree.py:
```diff
@@ -0,0 +1,87 @@
+        self.pid_to_guids_map = pid_to_guids_map if pid_to_guids_map else dict()
+        self.guid_to_processnode_map = guid_to_processnode_map if guid_to_processnode_map else dict()
+
+    async def add_processnode(self, guid, pid, link, parent_guid):
```
Added None as default

garunagiri commented on app/parsers/processguids.py:
```diff
@@ -24,19 +24,24 @@ def parse(self, blob):
     @staticmethod
     def parse_parentguid(blob):
-        return re.findall(r'\bParentProcessGuid: {(.*)}', blob, re.IGNORECASE)
+        return re.findall(r'\bParentProcessGuid:\W+{(.*)}', blob, re.IGNORECASE)
```
Added None as default

garunagiri replied:
I found that there were cases where that particular regex didn't work as intended. As that was the only one that was relevant to my use case, I didn't actually test the others

garunagiri commented:
Switched to using unique ProcessTrees per host (Chris' suggestion), and added descriptive text as docstrings for help understanding the contents (Henry's suggestion)
